File: #2CTinGĐ1619/CST1617/WEEK12/VQ12/HINH.pas
Uses Math;
CONST FI='HINH.INP';
      FO='HINH.OUT';
      INF=TRUNC(10E8);
TYPE LIST= ARRAY[0..45,0..45] OF BOOLEAN;
VAR N,M,TPLT: LONGINT;
    D: ARRAY [0..45,0..45] OF LONGINT;
    TD: ARRAY [1..4] OF LONGINT =(0,0,-1,1);
    TC: ARRAY [1..4] OF LONGINT =(-1,1,0,0);
    MINX,MINY,MAXY,MAXX,Area,Res: ARRAY [0..45*45] OF LONGINT;
    FREE: ARRAY [0..45*45] OF BOOLEAN;
    X,Y: LIST;
    A: ARRAY [0..45,0..45] OF CHAR;
    F: TEXT;

Procedure Init;
 Begin
 Assign(f,fi); Reset(F);
 Readln(f,n,m);
 For n:=1 to n do begin
  For m:=1 to m do Read(f,a[n,m]);
  Readln(F);
  End;
 Close(F);
 End;

Procedure DFS(x,y: Longint);
 Var u,v,i: Longint;
 Begin
 D[x,y]:=TPLT; Inc(Area[TPLT]);
 Maxx[TPLT]:=Max(x,Maxx[TPLT]);
 Minx[TPLT]:=Min(x,Minx[TPLT]);
 Maxy[TPLT]:=Max(y,Maxy[TPLT]);
 Miny[TPLT]:=Min(y,Miny[TPLT]);
 For i:=1 to 4 do
  Begin
  u:=x+td[i]; v:=y+tc[i];
  If D[u,v]=0 Then
   If A[u,v]=A[x,y] Then DFS(u,v);
  End;
 End;

Procedure CodeTrau;
 Var i,j: Longint;
 Begin
 TPLT:=0;
 Fillchar(Maxx,Sizeof(Maxx),0);
 Maxy:=Maxx;
 Fillchar(D,Sizeof(D),0);
 For i:=1 to n do
  For j:=1 to m do
   If D[i,j]=0 Then
    Begin
    Inc(TPLT);
    Minx[TPLT]:=inf;
    Miny[TPLT]:=inf;
    DFS(i,j);
    End;
 Writeln(f,TPLT);
 For i:=1 to n do Begin
  For j:=1 to m do
   Write(f,D[i,j],' ');
  Writeln(F);
  end;
 End;

Procedure XayMang(X: Longint;Var A:List);
 Var i,j: Longint;
 Begin
 Fillchar(A,sizeof(A),False);
 For i:=Minx[x] to Maxx[x] do
  For j:=Miny[x] to Maxy[x] do
   if (D[i,j]=x) then a[i-Minx[x]+1,j-Miny[x]+1]:=true;
 Maxx[x]:=Maxx[x]-Minx[x]+1;
 Maxy[x]:=Maxy[x]-Miny[x]+1;
 End;

Function Same(A,B: List): Boolean;
 Var i,j: Longint;
 Begin
 For i:=1 to n do
  For j:=1 to m do
   IF A[i,j]<>B[i,j] Then Exit(False);
 Exit(True);
 End;

Procedure Xoay(x: Longint;Var A: List);
 Var B: List;
     i,j,tmp: Longint;
 Begin
 Fillchar(B,Sizeof(B),False);
 for i:=1 to Maxx[x] do
  For j:=1 to Maxy[x] do B[Maxy[x]-j+1,i]:=A[i,j];
 Fillchar(A,sizeof(A),False);
 A:=B;
 Tmp:=Maxx[x]; Maxx[x]:=Maxy[x]; Maxy[x]:=Tmp;
 End;

Procedure CungAC;
 Var TPLT2,i,j,k: Longint;
 Begin
 TPLT2:=0;
 Fillchar(Free,sizeof(Free),True);
 For i:=1 to TPLT-1 do
  If Free[i] Then
   Begin
   Inc(TPLT2); Free[i]:=False;
   XayMang(i,X);
   Res[i]:=TPLT2;
   For j:=i+1 to TPLT do
    IF Area[i]=Area[j] Then Begin
     XayMang(j,Y);
     If Same(X,Y) Then Begin Free[j]:=False; Res[j]:=TPLT2; End
      Else
      For k:=1 to 3 do
       Begin
       Xoay(i,X);
       If Same(X,Y) Then Begin Free[j]:=False; Res[j]:=TPLT2; Break; End;
       End;
     End;
   End;
 Writeln(f,TPLT2);
 For i:=1 to n do Begin
  For j:=1 to m do Write(F,Res[D[i,j]],' ');
  Writeln(F);
  End;
 For i:=1 to TPLT do Write(f,Area[i],' ');
 End;

Begin
 Init;
 Assign(f,fo); Rewrite(F);
 CodeTrau;
 CungAC;
 Close(F);
End.

File: #2CTinGĐ1619/CST1617/WEEK12/VQ12/hinh.cpp
#include <iostream>
#include <cstdio>

#define maxN 45
#define maxn maxN*maxN

using namespace std;
int res[maxn],b[maxN][maxN],n,m,cnt=0,dd[]={1,-1,0,0},cc[]={0,0,1,-1},area[maxn],Minx[maxn],Miny[maxn],Maxx[maxn],Maxy[maxn],MAXX[maxn],MAXY[maxn];
char a[maxN][maxN]; bool visited[maxN][maxN],VISITED[maxn],SHAPE1[maxN][maxN],SHAPE2[maxN][maxN];

void DFS(int x,int y)
{
    b[x][y]=cnt;
    area[cnt]++;
    Minx[cnt]=min(Minx[cnt],x);
    Miny[cnt]=min(Miny[cnt],y);
    Maxx[cnt]=max(Maxx[cnt],x);
    Maxy[cnt]=max(Maxy[cnt],y);
    for (int i=0; i<4; i++)
    {
        int u=x+dd[i],v=y+cc[i];
        if (!visited[u][v] && a[u][v]==a[x][y])
        {
            visited[u][v]=true;
            DFS(u,v);
        }
    }
}

void yolo1()
{
    for (int i=1; i<=m; i++) for (int j=1; j<=n; j++) if (!visited[i][j])
    {
        cnt++;
        Minx[cnt]=Miny[cnt]=long(1e9);
        visited[i][j]=true;
        DFS(i,j);
    }
    cout<<cnt<<'\n';
    for (int i=1; i<=m; i++)
    {
        for (int j=1; j<=n; j++) cout<<b[i][j]<<" ";
        cout<<'\n';
    }
}

void fillchar(bool a[][maxN])
{
    for (int i=1; i<=40; i++) for (int j=1; j<=40; j++) a[i][j]=0;
}

void makeshape(int x,bool a[][maxN])
{
    fillchar(a);
    for (int i=Minx[x]; i<=Maxx[x]; i++) for (int j=Miny[x]; j<=Maxy[x]; j++) if (b[i][j]==x) a[i-Minx[x]+1][j-Miny[x]+1]=true;
    MAXX[x]=Maxx[x]-Minx[x]+1;
    MAXY[x]=Maxy[x]-Miny[x]+1;
}

bool same(bool a[][maxN],bool b[][maxN])
{
    for (int i=1; i<=m; i++) for (int j=1; j<=n; j++) if (a[i][j]!=b[i][j]) return false;
    return true;
}

void spinshape(bool a[][maxN],int x)
{
    bool b[maxN][maxN];
    for (int i=1; i<=MAXX[x]; i++) for (int j=1; j<=MAXY[x]; j++) b[MAXY[x]-j+1][i]=a[i][j];
    fillchar(a);
    for (int i=1; i<=MAXY[x]; i++) for (int j=1; j<=MAXX[x]; j++) a[i][j]=b[i][j];
    swap(MAXX[x],MAXY[x]);
}

void yolo2()
{
    int cntres=0;
    for (int shape1=1; shape1<cnt; shape1++) if (!VISITED[shape1])
    {
        makeshape(shape1,SHAPE1);
        cntres++;
        res[shape1]=cntres;
        for (int shape2=shape1+1; shape2<=cnt; shape2++) if (area[shape1]==area[shape2])
        {
            makeshape(shape2,SHAPE2);
            if (same(SHAPE1,SHAPE2))
            {
                VISITED[shape2]=true;
                res[shape2]=cntres;
            }
            else
            {
                for (int spin=1; spin<4; spin++)
                {
                    spinshape(SHAPE1,shape1);
                    if (same(SHAPE1,SHAPE2))
                    {
                        VISITED[shape2]=true;
                        res[shape2]=cntres;
                        break;
                    }
                }
            }
        }
    }
    cout<<cntres<<'\n';
    for (int i=1; i<=m; i++)
    {
        for (int j=1; j<=n; j++) cout<<res[b[i][j]]<<" ";
        cout<<'\n';
    }
    for (int i=1; i<=cnt; i++) cout<<area[i]<<' ';
}

int main()
{
    freopen("hinh.inp","r",stdin);
    freopen("hinh.out","w",stdout);
    scanf("%d%d\n",&m,&n);
    for (int i=1; i<=m; i++) scanf("%s",a[i]+1);
    yolo1();
    yolo2();
    return 0;
}
